OEP implementation

HassanJaveed84 opened the pull request:
Not sure how to deal with the 'mysql connector python' installation from a generic url in extra.txt, as pip compile does not support installations from generic urls.

HassanJaveed84 commented on scripts/post-pip-compile.sh:
```diff
@@ -0,0 +1,35 @@
+function clean_file {
+    FILE_PATH=$1
+    TEMP_FILE=${FILE_PATH}.tmp
+	# If an editable VCS URL has a version number suffix, it was only editable for pip-compile's benefit;
```
Initially added to remove -e flag from extra.txt, however it is not currently part of pip compile because of the generic url we use to install 'mysql connector python'

HassanJaveed84 replied:
Also, for our git based packages(luigi,pyinstrument) we don't tag it with a version.

jmbowman replied:
For `mysql-connector-python`, you could comment it out in the input file and include a section in `post-pip-compile.sh` that's something like this:

```bash
# Add back the mysql-connector-python package; this is a workaround for
# https://github.com/jazzband/pip-tools/issues/93
if [[ "${FILE_PATH}" == "requirements/extra.txt" ]]; then
    echo "http://cdn.mysql.com/Downloads/Connector-Python/mysql-connector-python-1.2.2.zip" >> ${FILE_PATH}
fi
```

I'd add an explanation above the commented-out line along the lines of

```
# pip-compile can't deal with this, it's added back by util/post-pip-compile.sh
```

HassanJaveed84 replied:
@jmbowman Should we be tagging our git based installs or remove them from pip-compile ?
We have two forks specified in extra.in(luigi and pyinstrument), we do not have fixed versions for those on our forks and currently specify hash:
``git+https://github.com/edx/luigi.git@eb45bcc52243de11b2b16a81229ac584fe1e601b#egg=luigi``
``git+https://github.com/edx/pyinstrument.git@a35ff76df4c3d5ff9a2876d859303e33d895e78f#egg=pyinstrument``

Should extra.in be removed from pip-compile or should we add a version tag to the said repos and then specify it in extra.in ?

jmbowman replied:
Ideally they should be tagged, even if just to ensure that the commit in use doesn't vanish later due to a force push or such.  But they should just work with pip-compile as is, given that they're installed as editable (the "-e" prefix).  We still have several dependencies like this in edx-platform also.

brianhw replied:
Just for the record, the -e prefix was added in this PR, just to get them back through pip-compile.   When they were in extra.txt, they did not use the -e prefix.

HassanJaveed84 commented:
@jmbowman can you please review this when you get a chance ?

HassanJaveed84 commented on requirements/base.txt:
```diff
@@ -2,24 +2,24 @@
-#    pip-compile --output-file requirements/base.txt requirements/base.in
+#    make upgrade
 #
 --no-binary ansible
```
base.in defines this here:
https://github.com/edx/edx-analytics-pipeline/blob/master/requirements/base.in#L3
I am not familiar with why we have it, maybe @mulby can shine some light on it.
It would automatically get included when running ```make upgrade```.

mulby replied:
I don't remember why we have this in here. Sorry!

brianhw replied:
I added this back in January, 2017 to master, and patched to Ficus branch in April, 2017.   This was a workaround for a "bug" in the old version of Ansible we're using.  The workaround was included in master along with a number of other changes, unfortunately.  The Ficus patch was more stand-alone:  https://github.com/edx/edx-analytics-pipeline/commit/60deb5360e3d1d1f693f800eca662da18ce4089e.   The comment:  "Addresses issue with ansible/ansible#8875, by pip installing with --no-binary flag."  So original issue is:  https://github.com/ansible/ansible/issues/8875.  If and when we upgrade ansible here, we can probably also remove this flag then.

jmbowman replied:
Thanks for the background.  Can we either try upgrading ansible or put that context in comments in `base.in`?

brianhw replied:
I agree that a comment at least would be useful.  I actually never tried to upgrade the ansible.  It might also not be too hard, but would involve more testing.  On the other hand, with the upgrades already happening, this should involve some thorough testing anyway.

brianhw commented on requirements/pip-tools.txt:
```diff
@@ -0,0 +1,10 @@
+click==6.7                # via pip-tools
+first==2.0.1              # via pip-tools
+pip-tools==2.0.2
+six==1.11.0               # via pip-tools
```
Note that this installs 1.11, and I *think* that if 1.11 is left in the virtualenv when trying to run the pipeline code that something breaks.  Hence all the hoops to use "--upgrade-strategy only-if-needed" to avoid this happening (before you took the different tack of removing the upgrade and including the .txt files instead of the .in files).  Perhaps this needs a comment in the pip-tools.in file?   Or maybe a suggestion somewhere that the "make upgrade" be done in a separate virtualenv?

jmbowman replied:
Ok, then `six` should probably be pinned in `pip-tools.in`, with an explanation for why.

HassanJaveed84 replied:
Should we be pinning this to 1.10.0?

brianhw replied:
If pip-compile works with 1.10.0, then I would pin it with a comment.  But it would be better if it were possible to put six>=1.10.0 in the pip-tools.in file, and it came through in that form in the pip-tools.txt file.  Then it should the right thing if installed into a fresh virtualenv or into a requirements-loaded virtualenv.

brianhw commented on Makefile:
```diff
@@ -34,23 +34,31 @@ else
-	pip install -U -r requirements/extra.txt --no-cache-dir --upgrade-strategy only-if-needed
+	pip install -r requirements/pip.txt
+	pip install -r requirements/default.txt --no-cache-dir
+	pip install -r requirements/extra.txt --no-cache-dir
```
So, what do we lose by removing the -U flags?  Is there some scenario where the packages would not get updated to the specified versions?   Perhaps it's a vestige of a time before all of our packages were pinned, and now it makes no sense anyway?

HassanJaveed84 commented:
@brianhw do you think this looks good to go ?

HassanJaveed84 commented:
@brianhw @jmbowman 

- Moved vcs urls to default.in, also moved psycopg2 to default.in. 
- Removed post-pip-compile script.
- Mysql connector python remains in extra.txt which is not part of pip-compile. 
- Tested doc generation using: ```make docs```
